packages/dd-trace/src/appsec/reporter.js

```diff
@@ -225,6 +242,54 @@ function reportWafInit (wafVersion, rulesVersion, diagnosticsRules = {}, success
   incrementWafInitMetric(wafVersion, rulesVersion, success)
 }
 
+function logWafDiagnosticMessage (product, rcConfigId, configKey, message, level) {
+  telemetryLogCh.publish({
+    message,
+    level,
+    tags: {
+      log_type: `rc::${product.toLowerCase()}::diagnostic`,
+      appsec_config_key: configKey,
+      rc_config_id: rcConfigId
+    }
+  })
+}
+
+function reportSuccessfulWafUpdate (product, rcConfigId, diagnostics) {
+  for (const configKey of WAF_DIAGNOSTICS_CONFIG_KEYS_TO_REPORT) {
+    const configDiagnostics = diagnostics[configKey]
+    if (!configDiagnostics) continue
+
+    if (configDiagnostics.error) {
+      logWafDiagnosticMessage(product, rcConfigId, configKey, configDiagnostics.error, 'ERROR')
+      continue
+    }
+
+    if (configDiagnostics.errors) {
+      for (const [errorMessage, errorIds] of Object.entries(configDiagnostics.errors)) {
+        logWafDiagnosticMessage(
+          product,
+          rcConfigId,
+          configKey,
+          `"${errorMessage}": ${JSON.stringify(errorIds)}`,
+          'ERROR'
+        )
+      }
+    }
+
+    if (configDiagnostics.warnings) {
+      for (const [warningMessage, warningIds] of Object.entries(configDiagnostics.warnings)) {
+        logWafDiagnosticMessage(
+          product,
+          rcConfigId,
+          configKey,
+          `"${warningMessage}": ${JSON.stringify(warningIds)}`,
+          'WARN'
+        )
+      }
+    }
```

Comment:
Suggestion (I think this is already good as is, so it's up to you): this could be unified with a loop over `['errors', 'warnings']`

Reply:
I prefer the way it is currently